### PR TITLE
ton: init at 2023.01

### PR DESCRIPTION
**Pull request by misuzu**
###### Description of changes
https://ton.org/
https://github.com/ton-blockchain/ton

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [X] aarch64-linux
  - [X] x86_64-darwin
  - [X] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/blockchains/ton/default.nix [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  buildInputs = [
+    gperf
+    libmicrohttpd
+    openssl
+    readline
+    zlib
+  ];
+
+  meta = with lib; {
+    description = "The next gen network to unite all blockchains and the existing Internet";
```
Comment (prusnak) [inline pkgs/applications/blockchains/ton/default.nix:46]: This is ridiculous - please change to something that doesn't read like an ad and rather has some informational value.

File: pkgs/applications/blockchains/ton/default.nix [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    homepage = "https://ton.org/";
+    license = licenses.gpl2Only;
```
Comment (prusnak) [inline pkgs/applications/blockchains/ton/default.nix:48]: License is LGPLv2 not GPLv2

```suggestion
    license = licenses.lgpl2Only;
```

**Review by prusnak — CHANGES_REQUESTED** [review]

File: pkgs/applications/blockchains/ton/default.nix (hunk 2 above)
Comment (misuzu) [reply pkgs/applications/blockchains/ton/default.nix]: Fixed!

File: pkgs/applications/blockchains/ton/default.nix (hunk 1 above)
Comment (misuzu) [reply pkgs/applications/blockchains/ton/default.nix]: Is it better now?

File: pkgs/applications/blockchains/ton/default.nix [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    sha256 = "sha256-/eY/rPHQEoPjdbHmTpg2f8rjkg8tKAed4oF8KoMqz8k=";
+    fetchSubmodules = true;
+  };
+
+  postPatch = ''
+    # do not set TON_ARCH or it will use wrong -mcpu/-march
+    substituteInPlace CMakeLists.txt \
+      --replace 'set(TON_ARCH "native" CACHE STRING "Architecture, will be passed to -march=")' "" \
+      --replace 'set(TON_ARCH "apple-m1")' ""
+  '';
```
Comment (NickCao) [inline pkgs/applications/blockchains/ton/default.nix:30]: Isn't our cc-wrapper filter these flags out?

Comment (misuzu) [reply pkgs/applications/blockchains/ton/default.nix]: Yes, but not "apple-m1". Fixed.

**Review by NickCao — APPROVED** [review]

**Pull request merged**